subsys/bluetooth/services/hids_c.c

```diff
@@ -0,0 +1,864 @@
+	rep->notify_params.value = BT_GATT_CCC_NOTIFY;
+	rep->notify_params.value_handle = rep->handlers.val;
+	rep->notify_params.ccc_handle = rep->handlers.ccc;
+
+	err = bt_gatt_subscribe(hids_c->conn, &rep->notify_params);
+	if (err) {
+		LOG_ERR("Report notification subscribe error: %d.", err);
+		rep->notify_cb = NULL;
+		return err;
+	}
+	LOG_DBG("Report subscribed.");
+	return err;
+}
+
+int bt_gatt_hids_c_rep_unsubscribe(struct bt_gatt_hids_c *hids_c,
+				   struct bt_gatt_hids_c_rep_info *rep)
+{
+	if (!hids_c || !rep) {
+		return -EINVAL;
+	}
+	return bt_gatt_unsubscribe(hids_c->conn, &rep->notify_params);
+}
+
+/**
+ * @brief Process map read
+ *
+ * Internal function to process report map read and pass it further.
+ *
+ * @param conn   Connection handler.
+ * @param err    Read ATT error code.
+ * @param params Notification parameters structure - the pointer
+ *               to the structure provided to read function.
+ * @param data   Pointer to the data buffer.
+ * @param length The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP     Stop notification
+ * @retval BT_GATT_ITER_CONTINUE Continue notification
+ */
+static u8_t map_read_process(struct bt_conn *conn, u8_t err,
+			     struct bt_gatt_read_params *params,
+			     const void *data, u16_t length)
+{
+	struct bt_gatt_hids_c *hids_c;
+	size_t offset;
+
+	hids_c = CONTAINER_OF(params,
+			      struct bt_gatt_hids_c,
+			      read_params);
+	if (!hids_c->map_cb) {
+		LOG_ERR("No map read callback present");
+		return BT_GATT_ITER_STOP;
+	}
+
+	offset = hids_c->read_params.single.offset;
+	k_sem_give(&hids_c->read_params_sem);
+	hids_c->map_cb(hids_c, err, data, length, offset);
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_gatt_hids_c_map_read(struct bt_gatt_hids_c *hids_c,
+			    bt_gatt_hids_c_map_cb func,
+			    size_t offset,
+			    s32_t timeout)
+{
+	int err;
+
+	if (!hids_c || !func) {
+
+		return -EINVAL;
+	}
+	err = k_sem_take(&hids_c->read_params_sem, timeout);
+	if (err) {
+		return err;
+	}
+	hids_c->map_cb = func;
+	hids_c->read_params.func = map_read_process;
+	hids_c->read_params.handle_count  = 1;
+	hids_c->read_params.single.handle = hids_c->handlers.rep_map;
+	hids_c->read_params.single.offset = offset;
+	err = bt_gatt_read(hids_c->conn, &(hids_c->read_params));
+	if (err) {
+		hids_c->map_cb = NULL;
+		k_sem_give(&hids_c->read_params_sem);
+		return err;
+	}
+	return 0;
+}
+
+/**
+ * @brief Protocol mode update process
+ *
+ * Function that process protocol mode update and updates
+ * the HIDS client object.
+ *
+ * @param conn   Connection handler.
+ * @param err    Read ATT error code.
+ * @param params Notification parameters structure - the pointer
+ *               to the structure provided to read function.
+ * @param data   Pointer to the data buffer.
+ * @param length The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP     Stop notification
+ * @retval BT_GATT_ITER_CONTINUE Continue notification
```

Comment:
Not returned by this function.

Reply:
This is callback function and documentation describes the reaction of the caller.